accepted/future-releases/nnbd/feature-specification.md

```diff
@@ -263,6 +268,157 @@ The **flatten** function is modified as follows:
 and for all `R`, if `T <: Future<R>` then `S <: R`; then **flatten**(`T`) = `S`
   - otherwise **flatten**(`T`) = `T`
 
+### The future value type of an asynchronous generator
+
+_We specify a concept which corresponds to the type of object which is
+delivered by an `async` function, that is, the type of object which is used to
+complete the future which is returned from such a function when it suspends
+for the first time._
+
+Let _f_ be an asynchronous non-generator function. We define the notion of the
+**future value type** of _f_ as follows: If, for some `U`, the declared return
+type `T` of _f_ is `Future<U>`, `FutureOr<U>`, `Future<U>?`, or `FutureOr<U>?`,
+the future value type of _f_ is `U`.
+Otherwise, if `T` is `void` then the future value type of _f_ is `void`.
+Otherwise, the future value type of _f_ is `Object?`.
+
+### Return statements
+
+The static analysis of return statements is changed in the following
+way, where `$T$` is the declared return type and `$S$` is the static type of
+the expression `e`:
+
+At [this location](https://github.com/dart-lang/language/blob/65b8267be0ebb9b3f0849e2061e6132021a4827d/specification/dartLangSpec.tex#L15477)
+about synchronous non-generator functions, the text is changed as follows:
+
+```
+It is a compile-time error if $s$ is \code{\RETURN{} $e$;},
+$T$ is neither \VOID{} nor \DYNAMIC,
+and $S$ is \VOID.
+```
+
+_Comparing to Dart before null-safety, this means that it is no longer allowed
+to "return void to null" in a regular function._
+
+At [this location](https://github.com/dart-lang/language/blob/65b8267be0ebb9b3f0849e2061e6132021a4827d/specification/dartLangSpec.tex#L15525)
+about an asynchronous non-generator function with future value type `$T_v$`,
+the text is changed as follows:
+
+```
+It is a compile-time error if $s$ is \code{\RETURN{} $e$;},
+$T_v$ is neither \VOID{} nor \DYNAMIC,
+and \flatten{S} is \VOID{}.
```

Comment:
I know that @johnniwinther considers normalization "everywhere" to be disruptive, so that's already an important thing to take into account. I think one of the reasons for this is performance, but it's better that @johnniwinther comments on this as he sees fit.

This means that it may be a good choice to apply normalization rather sparsely. An example could be that some people want to use the type `FutureOr<void>` to indicate that a value may be a future whose completion should be ignored, or it might be "something else" which should be ignored. Of course, that type is questionable in its own right, but we decided to allow it, and it does get special treatment: It is allowed to perform a type test on an expression with that type, so it is not treated exactly like an expression of type `void`.

Reply:
I changed the rule to raise the error when `$S$` is `void`, `void?`, or `void*`.

It is possible that we won't mention anything involving legacy types in the language specification, but I think it makes sense to include `void*` here, and then it is easy to delete it when the language specification is updated, if that's what we do.

Reply:
The main problem with normalization is when it leaks into the types that we present to the user. If the user has written a type one way but we report problems pertaining to that type in another way it is very confusing. Like

```Dart
Error: Bar isn't a valid type:
  Foo foo = ....
  ^^^
```
For that reason I'd like to avoid normalization in cases where it is not needed for the correct result, i.e. in cases where the user provided type suffices, and only use normalization why we need to synthesize a result.

This is for instance the strategy we use when computing the canonical supertypes:
```Dart
class A<T> {}
class B implements A<FutureOr<void>> {} // B implements A<FutureOr<void>>
class C implements A<dynamic> {} // C implements A<dynamic>
class D implements B, C {} // D implements A<Object?>
```
where we only apply `NNBD_TOP_MERGE(NORM(A<FutureOr<void>>),NORM(A<dynamic>)) = A<Object?>` because we need to for a well-defined supertype relation.